File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    
    const visited = Array.from(Array(n), () => Array(m).fill(0));
    visited[0][0] = 1
    
    const q = [[0, 0]];
    
    const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];
    while (q.length > 0) {
        const [x, y] = q.shift();
        if (x === n-1 && y === m-1) {
            return visited[x][y];
        }
        for (let [dx, dy] of directions) {
            const nx = x + dx;
            const ny = y + dy;
            if (
                n > nx &&
                nx >= 0 &&
                m > ny &&
                ny >= 0 &&
                maps[nx][ny] === 1 &&
                visited[nx][ny] === 0
            ) {
                q.push([nx, ny])
                visited[nx][ny] = visited[x][y] + 1
            }
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
def solution(maps):
    answer = 0
    N, M = len(maps), len(maps[0])
    q = [(0, 0)]
    visited = [[0 for _ in range(M)] for _ in range(N)]
    visited[0][0] = 1
    while q:
        x, y = q.pop(0)
        for dx, dy in (1, 0), (-1, 0), (0, 1), (0, -1):
            nx, ny = x + dx, y + dy
            if N > nx >= 0 and M > ny >= 0 and not visited[nx][ny] and maps[nx][ny]:
                if nx == N-1 and ny == M-1:
                    return visited[x][y] + 1
                visited[nx][ny] = visited[x][y] + 1
                q.append((nx, ny))
    return -1